server/resp/src/main/java/org/infinispan/server/resp/serialization/Resp3SerializerRegistry.java

```diff
@@ -0,0 +1,127 @@
+      serialize(value, alloc, candidates.get());
+   }
+
+   static void serialize(Object value, ByteBufPool alloc, ResponseSerializer<?>[] candidates) {
+      // The first check must always be for the null serializer.
+      // This removes all null checks on every other serializer.
+      if (PrimitiveSerializer.NullSerializer.INSTANCE.test(value)) {
+         PrimitiveSerializer.NullSerializer.INSTANCE.accept(null, alloc);
+         return;
+      }
+
+      for (ResponseSerializer<?> serializer : candidates) {
+         if (serializer.test(value)) {
+            serialize(serializer, value, alloc);
+            return;
+         }
+      }
+
+      throw new IllegalStateException("Serializer unknown: " + value.getClass());
+   }
+
+   /**
+    * Serializes the given object only with the provided serializer.
+    *
+    * @param value The object to serialize in RESP3 format.
+    * @param alloc Buffer pool for allocation.
+    * @param candidate The candidate serializer to handle the object.
+    * @throws IllegalStateException: when the candidate does not accept the provided value.
+    */
+   static void serialize(Object value, ByteBufPool alloc, ResponseSerializer<?> candidate) {
+      // The first check must always be for the null serializer.
+      // This removes all null checks on every other serializer.
+      if (PrimitiveSerializer.NullSerializer.INSTANCE.test(value)) {
+         PrimitiveSerializer.NullSerializer.INSTANCE.accept(null, alloc);
+         return;
+      }
+
+      if (!candidate.test(value))
+         throw new IllegalStateException("Serializer not handling: " + value.getClass());
+
+      serialize(candidate, value, alloc);
+   }
+
+   static <H extends SerializationHint> void serialize(Object value, ByteBufPool alloc,
+                                                       NestedResponseSerializer<?, H> candidate, H hint) {
+
+      // The first check must always be for the null serializer.
+      // This removes all null checks on every other serializer.
+      if (PrimitiveSerializer.NullSerializer.INSTANCE.test(value)) {
+         PrimitiveSerializer.NullSerializer.INSTANCE.accept(null, alloc);
+         return;
+      }
+
+      if (!candidate.test(value))
+         throw new IllegalStateException("Serializer not handling: " + value.getClass());
+
+      @SuppressWarnings("unchecked")
+      NestedResponseSerializer<Object, H> nrs = (NestedResponseSerializer<Object, H>) candidate;
+      nrs.accept(value, alloc, hint);
+   }
+
+   private static void serialize(ResponseSerializer<?> serializer, Object value, ByteBufPool alloc) {
+      @SuppressWarnings("unchecked")
+      ResponseSerializer<Object> s = (ResponseSerializer<Object>) serializer;
+      s.accept(value, alloc);
+   }
+
+   /**
+    * Register a custom serializer.
+    *
+    * @param serializer The custom serializer.
+    */
+   public static synchronized void addCustomSerializer(JavaObjectSerializer<?> serializer) {
```

Comment:
I question if this is really worth it. At worst we could just add all these infrequently used serializers to the end of the array. I think it would be more beneficial to do this and have the `candidates` above just be a static final ResponseSerializer<?>[] instead. At worst case it just means it would take a few iterations more before we throw an exception, which shouldn't be happening anyways. But the inverse is we remove the volatile read from the AtomicReference and also can remove the Collection.

Reply:
Managed to remove all this. Now, we keep only the ResponseSerializer<?>[] array statically initiated.